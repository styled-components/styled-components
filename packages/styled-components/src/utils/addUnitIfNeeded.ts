import unitless from '@emotion/unitless';

// Taken from https://github.com/facebook/react/blob/b87aabdfe1b7461e7331abb3601d9e6bb27544bc/packages/react-dom/src/shared/dangerousStyleValue.js
export default function addUnitIfNeeded(name: string, value: any) {
  // https://github.com/amilajack/eslint-plugin-flowtype-errors/issues/133
  if (value == null || typeof value === 'boolean' || value === '') {
    return '';
  }

  if (typeof value === 'number' && value !== 0 && !(name in unitless)) {
    return `${value}px`; // Presumes implicit 'px' suffix for unitless numbers
  }

  return String(value).trim();
}
