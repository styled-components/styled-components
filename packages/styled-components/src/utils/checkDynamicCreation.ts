import { useRef } from 'react';

const invalidHookCallRe = /invalid hook call/i;
const seen = new Set();

export const checkDynamicCreation = (displayName: string, componentId?: string) => {
  if (process.env.NODE_ENV !== 'production') {
    const parsedIdString = componentId ? ` with the id of "${componentId}"` : '';
    const message =
      `The component ${displayName}${parsedIdString} has been created dynamically.\n` +
      "You may see this warning because you've called styled inside another component.\n" +
      'To resolve this only create new StyledComponents outside of any render method and function component.';

    // If a hook is called outside of a component:
    // React 17 and earlier throw an error
    // React 18 and above use console.error

    const originalConsoleError = console.error
    try {
      let didNotCallInvalidHook = true
      console.error = (consoleErrorMessage, ...consoleErrorArgs) => {
      // The error here is expected, since we're expecting anything that uses `checkDynamicCreation` to
      // be called outside of a React component.
        if (invalidHookCallRe.test(consoleErrorMessage)) {
          didNotCallInvalidHook = false
          // This shouldn't happen, but resets `warningSeen` if we had this error happen intermittently
          seen.delete(message);
        } else {
          originalConsoleError(consoleErrorMessage, ...consoleErrorArgs);
        }
      }
      // We purposefully call `useRef` outside of a component and expect it to throw
      // If it doesn't, then we're inside another component.
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useRef();

      if (didNotCallInvalidHook && !seen.has(message)) {
        // eslint-disable-next-line no-console
        console.warn(message);
        seen.add(message);
      }
    } catch (error) {
      // The error here is expected, since we're expecting anything that uses `checkDynamicCreation` to
      // be called outside of a React component.
      if (invalidHookCallRe.test(error.message)) {
        // This shouldn't happen, but resets `warningSeen` if we had this error happen intermittently
        seen.delete(message);
      }
    } finally {
      console.error = originalConsoleError;
    }
  }
};
