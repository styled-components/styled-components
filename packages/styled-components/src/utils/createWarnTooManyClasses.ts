import { ExtensibleObject } from '../types';

export const LIMIT = 200;

export default (displayName: string, componentId: string) => {
  let generatedClasses: ExtensibleObject = {};
  let warningSeen = false;

  return (className: string) => {
    if (!warningSeen) {
      generatedClasses[className] = true;
      if (Object.keys(generatedClasses).length >= LIMIT) {
        // Unable to find latestRule in test environment.

        /* eslint-disable no-console, prefer-template */
        const parsedIdString = componentId ? ` with the id of "${componentId}"` : '';

        console.warn(
          `Over ${LIMIT} classes were generated for component ${displayName}${parsedIdString}.\n` +
            'Consider using the attrs method, together with a style object for frequently changed styles.\n' +
            'Example:\n' +
            '  const Component = styled.div.attrs(props => ({\n' +
            '    style: {\n' +
            '      background: props.background,\n' +
            '    },\n' +
            '  }))`width: 100%;`\n\n' +
            '  <Component />'
        );
        warningSeen = true;
        generatedClasses = {};
      }
    }
  };
};
