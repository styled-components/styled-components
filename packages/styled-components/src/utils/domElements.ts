// Thanks to ReactDOMFactories for this handy list!

export default [
  'a',
  'abbr',
  'address',
  'area',
  'article',
  'aside',
  'audio',
  'b',
  'base',
  'bdi',
  'bdo',
  'big',
  'blockquote',
  'body',
  'br',
  'button',
  'canvas',
  'caption',
  'cite',
  'code',
  'col',
  'colgroup',
  'data',
  'datalist',
  'dd',
  'del',
  'details',
  'dfn',
  'dialog',
  'div',
  'dl',
  'dt',
  'em',
  'embed',
  'fieldset',
  'figcaption',
  'figure',
  'footer',
  'form',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'head',
  'header',
  'hgroup',
  'hr',
  'html',
  'i',
  'iframe',
  'img',
  'input',
  'ins',
  'kbd',
  'keygen',
  'label',
  'legend',
  'li',
  'link',
  'main',
  'map',
  'mark',
  'menu',
  'menuitem',
  'meta',
  'meter',
  'nav',
  'noscript',
  'object',
  'ol',
  'optgroup',
  'option',
  'output',
  'p',
  'param',
  'picture',
  'pre',
  'progress',
  'q',
  'rp',
  'rt',
  'ruby',
  's',
  'samp',
  'script',
  'section',
  'select',
  'small',
  'source',
  'span',
  'strong',
  'style',
  'sub',
  'summary',
  'sup',
  'table',
  'tbody',
  'td',
  'textarea',
  'tfoot',
  'th',
  'thead',
  'time',
  'title',
  'tr',
  'track',
  'u',
  'ul',
  'var',
  'video',
  'wbr', // SVG
  'circle',
  'clipPath',
  'defs',
  'ellipse',
  'foreignObject',
  'g',
  'image',
  'line',
  'linearGradient',
  'marker',
  'mask',
  'path',
  'pattern',
  'polygon',
  'polyline',
  'radialGradient',
  'rect',
  'stop',
  'svg',
  'text',
  'tspan',
] as const;
