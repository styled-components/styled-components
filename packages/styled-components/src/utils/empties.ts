import { ExtensibleObject } from '../types';

export const EMPTY_ARRAY = (Object.freeze([]) as unknown) as Readonly<any[]>;
export const EMPTY_OBJECT = Object.freeze({}) as Readonly<ExtensibleObject>;
