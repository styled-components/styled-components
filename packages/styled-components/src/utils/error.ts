import { ExtensibleObject } from '../types';
import errorMap from './errors';

const ERRORS: ExtensibleObject = process.env.NODE_ENV !== 'production' ? errorMap : {};

/**
 * super basic version of sprintf
 */
function format(...args: [string, ...any]) {
  let a = args[0];
  const b = [];

  for (let c = 1, len = args.length; c < len; c += 1) {
    b.push(args[c]);
  }

  b.forEach(d => {
    a = a.replace(/%[a-z]/, d);
  });

  return a;
}

/**
 * Create an error file out of errors.md for development and a simple web link to the full errors
 * in production mode.
 */
export default function throwStyledComponentsError(
  code: string | number,
  ...interpolations: any[]
) {
  if (process.env.NODE_ENV === 'production') {
    return new Error(
      `An error occurred. See https://git.io/JUIaE#${code} for more information.${
        interpolations.length > 0 ? ` Args: ${interpolations.join(', ')}` : ''
      }`
    );
  } else {
    return new Error(format(ERRORS[code], ...interpolations).trim());
  }
}
