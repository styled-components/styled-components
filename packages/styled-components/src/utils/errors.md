**DO NOT REMOVE ERRORS OR MAKE NEW ONES OUT OF SEQUENCE, ALWAYS ADD TO THE END.**

## 1

Cannot create styled-component for component: %s.

## 2

Can't collect styles once you've consumed a `ServerStyleSheet`'s styles! `ServerStyleSheet` is a one off instance for each server-side render cycle.

- Are you trying to reuse it across renders?
- Are you accidentally calling collectStyles twice?

## 3

Streaming SSR is only supported in a Node.js environment; Please do not try to call this method in the browser.

## 4

The `StyleSheetManager` expects a valid target or sheet prop!

- Does this error occur on the client and is your target falsy?
- Does this error occur on the server and is the sheet falsy?

## 5

The clone method cannot be used on the client!

- Are you running in a client-like environment on the server?
- Are you trying to run SSR on the client?

## 6

Trying to insert a new style tag, but the given Node is unmounted!

- Are you using a custom target that isn't mounted?
- Does your document not have a valid head element?
- Have you accidentally removed a style tag manually?

## 7

ThemeProvider: Please return an object from your "theme" prop function, e.g.

```js
theme={() => ({})}
```

## 8

ThemeProvider: Please make your "theme" prop an object.

## 9

Missing document `<head>`

## 10

Cannot find a StyleSheet instance. Usually this happens if there are multiple copies of styled-components loaded at once. Check out this issue for how to troubleshoot and fix the common cases where this situation can happen: https://github.com/styled-components/styled-components/issues/1941#issuecomment-417862021

## 11

_This error was replaced with a dev-time warning, it will be deleted for v4 final._ [createGlobalStyle] received children which will not be rendered. Please use the component without passing children elements.

## 12

It seems you are interpolating a keyframe declaration (%s) into an untagged string. This was supported in styled-components v3, but is not longer supported in v4 as keyframes are now injected on-demand. Please wrap your string in the css\`\` helper which ensures the styles are injected correctly. See https://www.styled-components.com/docs/api#css

## 13

%s is not a styled component and cannot be referred to via component selector. See https://www.styled-components.com/docs/advanced#referring-to-other-components for more details.

## 14

ThemeProvider: "theme" prop is required.

## 15

A stylis plugin has been supplied that is not named. We need a name for each plugin to be able to prevent styling collisions between different stylis configurations within the same app. Before you pass your plugin to `<StyleSheetManager stylisPlugins={[]}>`, please make sure each plugin is uniquely-named, e.g.

```js
Object.defineProperty(importedPlugin, 'name', { value: 'some-unique-name' });
```

## 16

Reached the limit of how many styled components may be created at group %s.
You may only create up to 1,073,741,824 components. If you're creating components dynamically,
as for instance in your render method then you may be running into this limitation.

## 17

CSSStyleSheet could not be found on HTMLStyleElement.
Has styled-components' style tag been unmounted or altered by another script?
