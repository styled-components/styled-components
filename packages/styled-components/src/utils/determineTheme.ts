import { ExtensibleObject } from '../types';
import { EMPTY_OBJECT } from './empties';

export default function determineTheme(
  props: ExtensibleObject,
  providedTheme: any,
  defaultProps: any = EMPTY_OBJECT
) {
  return (props.theme !== defaultProps.theme && props.theme) || providedTheme || defaultProps.theme;
}
