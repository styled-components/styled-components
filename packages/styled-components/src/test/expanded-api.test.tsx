import React from 'react';
import TestRenderer from 'react-test-renderer';
import { getRenderedCSS, resetStyled } from './utils';

// Disable isStaticRules optimisation since we're not
// testing for ComponentStyle specifics here
jest.mock('../utils/isStaticRules', () => () => false);

let styled: ReturnType<typeof resetStyled>;

describe('expanded api', () => {
  /**
   * Make sure the setup is the same for every test
   */
  beforeEach(() => {
    styled = resetStyled();
  });

  describe('displayName', () => {
    it('should be auto-generated if none passed', () => {
      const Comp = styled.div``;
      expect(Comp.displayName).toBe('styled.div');
    });

    it('should be attached if supplied', () => {
      const Comp = styled.div.withConfig({ displayName: 'Comp' })``;
      expect(Comp.displayName).toBe('Comp');
    });
  });

  describe('componentId', () => {
    it('should be generated as "sc" + hash', () => {
      const Comp = styled.div``;
      const Comp2 = styled.div``;
      expect(Comp.styledComponentId).toBe('sc-a');
      expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
      expect(Comp2.styledComponentId).toBe('sc-b');
      expect(TestRenderer.create(<Comp2 />)).toMatchSnapshot();
    });

    it('should be generated from displayName + hash', () => {
      const Comp = styled.div.withConfig({ displayName: 'Comp' })``;
      const Comp2 = styled.div.withConfig({ displayName: 'Comp2' })``;
      expect(Comp.styledComponentId).toBe('Comp-a');
      expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
      expect(Comp2.styledComponentId).toBe('Comp2-b');
      expect(TestRenderer.create(<Comp2 />)).toMatchSnapshot();
    });

    it('should be attached if passed in', () => {
      const Comp = styled.div.withConfig({ componentId: 'LOLOMG' })``;
      const Comp2 = styled.div.withConfig({ componentId: 'OMGLOL' })``;
      expect(Comp.styledComponentId).toBe('LOLOMG');
      expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
      expect(Comp2.styledComponentId).toBe('OMGLOL');
      expect(TestRenderer.create(<Comp2 />)).toMatchSnapshot();
    });

    it('should be combined with displayName if both passed in', () => {
      const Comp = styled.div.withConfig({
        displayName: 'Comp',
        componentId: 'LOLOMG',
      })``;
      const Comp2 = styled.div.withConfig({
        displayName: 'Comp2',
        componentId: 'OMGLOL',
      })``;
      expect(Comp.styledComponentId).toBe('Comp-LOLOMG');
      expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
      expect(Comp2.styledComponentId).toBe('Comp2-OMGLOL');
      expect(TestRenderer.create(<Comp2 />).toJSON()).toMatchSnapshot();
    });

    it('should work with `.withComponent`', () => {
      const Dummy: React.FC<any> = props => <div {...props} />;
      const Comp = styled.div.withConfig({
        displayName: 'Comp',
        componentId: 'OMGLOL',
      })``.withComponent('h1');
      const Comp2 = styled.div.withConfig({
        displayName: 'Comp2',
        componentId: 'OMFG',
      })``.withComponent(Dummy);
      expect(Comp.styledComponentId).toBe('Comp-OMGLOL-h1');
      expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
      expect(Comp2.styledComponentId).toBe('Comp2-OMFG-Dummy');
      expect(TestRenderer.create(<Comp2 />).toJSON()).toMatchSnapshot();
    });
  });

  describe('chaining', () => {
    it('should merge the options strings', () => {
      const Comp = styled.div
        .withConfig({ componentId: 'id-1' })
        .withConfig({ displayName: 'dn-2' })``;
      expect(Comp.displayName).toBe('dn-2');
      expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
    });

    it('should keep the last value passed in when merging', () => {
      const Comp = styled.div
        .withConfig({ displayName: 'dn-2', componentId: 'id-3' })
        .withConfig({ displayName: 'dn-5', componentId: 'id-4' })``;
      expect(Comp.displayName).toBe('dn-5');
      expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
    });
  });

  describe('"as" prop', () => {
    it('changes the rendered element type', () => {
      const Comp = styled.div`
        color: red;
      `;

      expect(TestRenderer.create(<Comp as="span" />).toJSON()).toMatchSnapshot();
    });

    it('changes the rendered element type when used with attrs', () => {
      const Comp = styled.div.attrs(() => ({
        as: 'header',
      }))`
        color: red;
      `;

      expect(TestRenderer.create(<Comp />).toJSON()).toMatchSnapshot();
    });

    it('prefers attrs over props', () => {
      const Comp = styled.div.attrs(() => ({
        as: 'header',
      }))`
        color: red;
      `;

      expect(TestRenderer.create(<Comp as="span" />).toJSON()).toMatchInlineSnapshot(`
        <header
          className="sc-a b"
        />
      `);
    });

    it('works with custom components', () => {
      const Override: React.FC<any> = props => <figure {...props} />;
      const Comp = styled.div`
        color: red;
      `;

      expect(TestRenderer.create(<Comp as={Override} />).toJSON()).toMatchSnapshot();
    });

    it('transfers all styles that have been applied', () => {
      const Comp = styled.div`
        background: blue;
        color: red;
      `;

      const Comp2 = styled(Comp)`
        color: green;
      `;

      const Comp3 = styled(Comp2)`
        text-align: center;
      `;

      expect(Comp.displayName).toMatchInlineSnapshot(`"styled.div"`);
      expect(Comp2.displayName).toMatchInlineSnapshot(`"Styled(styled.div)"`);
      expect(Comp3.displayName).toMatchInlineSnapshot(`"Styled(Styled(styled.div))"`);
      expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
        <div
          className="sc-a d"
        />
      `);
      expect(TestRenderer.create(<Comp2 />).toJSON()).toMatchInlineSnapshot(`
        <div
          className="sc-a sc-b d e"
        />
      `);
      expect(TestRenderer.create(<Comp3 as="span" />).toJSON()).toMatchInlineSnapshot(`
        <span
          className="sc-a sc-b sc-c d e f"
        />
      `);

      expect(getRenderedCSS()).toMatchInlineSnapshot(`
        ".d {
          background: blue;
          color: red;
        }
        .e {
          color: green;
        }
        .f {
          text-align: center;
        }"
      `);
    });
  });
});
