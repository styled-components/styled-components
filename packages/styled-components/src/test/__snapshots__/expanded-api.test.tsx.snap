// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expanded api "as" prop changes the rendered element type 1`] = `
<span
  className="sc-a b"
/>
`;

exports[`expanded api "as" prop changes the rendered element type when used with attrs 1`] = `
<header
  className="sc-a b"
/>
`;

exports[`expanded api "as" prop works with custom components 1`] = `
<figure
  className="sc-a b"
/>
`;

exports[`expanded api chaining should keep the last value passed in when merging 1`] = `
<div
  className="dn-5-id-4"
/>
`;

exports[`expanded api chaining should merge the options strings 1`] = `
<div
  className="dn-2-id-1"
/>
`;

exports[`expanded api componentId should be attached if passed in 1`] = `
<div
  className="LOLOMG"
/>
`;

exports[`expanded api componentId should be attached if passed in 2`] = `
<div
  className="OMGLOL"
/>
`;

exports[`expanded api componentId should be combined with displayName if both passed in 1`] = `
<div
  className="Comp-LOLOMG"
/>
`;

exports[`expanded api componentId should be combined with displayName if both passed in 2`] = `
<div
  className="Comp2-OMGLOL"
/>
`;

exports[`expanded api componentId should be generated as "sc" + hash 1`] = `
<div
  className="sc-a"
/>
`;

exports[`expanded api componentId should be generated as "sc" + hash 2`] = `
<div
  className="sc-b"
/>
`;

exports[`expanded api componentId should be generated from displayName + hash 1`] = `
<div
  className="Comp-a"
/>
`;

exports[`expanded api componentId should be generated from displayName + hash 2`] = `
<div
  className="Comp2-b"
/>
`;

exports[`expanded api componentId should work with \`.withComponent\` 1`] = `
<h1
  className="Comp-OMGLOL-h1"
/>
`;

exports[`expanded api componentId should work with \`.withComponent\` 2`] = `
<div
  className="Comp2-OMFG-Dummy"
/>
`;
